Content.Server/Ninja/Systems/NinjaSystem.cs

```diff
@@ -0,0 +1,572 @@
+    [Dependency] private readonly TraitorRuleSystem _traitorRule = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ActivateInWorldEvent>(OnActivate);
+        SubscribeLocalEvent<NinjaDrainComponent, DoAfterEvent>(OnDrainDoAfter);
+        SubscribeLocalEvent<NinjaDownloadComponent, DoAfterEvent>(OnDownloadDoAfter);
+        SubscribeLocalEvent<NinjaTerrorComponent, DoAfterEvent>(OnTerrorDoAfter);
+
+        // TODO: maybe have suit activation stuff
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, ContainerIsInsertingAttemptEvent>(OnSuitInsertAttempt);
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, TogglePhaseCloakEvent>(OnTogglePhaseCloakAction);
+        SubscribeLocalEvent<SpaceNinjaSuitComponent, RecallKatanaEvent>(OnRecallKatanaAction);
+
+        SubscribeLocalEvent<SpaceNinjaComponent, ComponentStartup>(OnNinjaStartup);
+        SubscribeLocalEvent<SpaceNinjaComponent, MindAddedMessage>(OnNinjaMindAdded);
+        SubscribeLocalEvent<SpaceNinjaComponent, AttackedEvent>(OnNinjaAttacked);
+
+        SubscribeLocalEvent<DoorComponent, DoorEmaggedEvent>(OnDoorEmagged);
+    }
+
+    public override void Update(float frameTime)
+    {
+        foreach (var ninja in EntityQuery<SpaceNinjaComponent>())
+        {
+            var uid = ninja.Owner;
+            UpdateNinja(uid, ninja, frameTime);
+        }
+    }
+
+    /// <summary>
+    /// Turns the player into a space ninja
+    /// </summary>
+    public void MakeNinja(Mind.Mind mind)
+    {
+        if (mind.OwnedEntity == null)
+            return;
+
+        // prevent double ninja'ing
+        var user = mind.OwnedEntity.Value;
+        if (HasComp<SpaceNinjaComponent>(user))
+            return;
+
+        AddComp<SpaceNinjaComponent>(user);
+        SetOutfitCommand.SetOutfit(user, "SpaceNinjaGear", EntityManager);
+        GreetNinja(mind);
+    }
+
+    protected override void NinjaEquippedGloves(EntityUid uid, SpaceNinjaGlovesComponent comp, EntityUid user, SpaceNinjaComponent ninja)
+    {
+        base.NinjaEquippedGloves(uid, comp, user, ninja);
+
+        if (TryComp<ActionsComponent>(user, out var actions))
+            _actions.AddAction(user, comp.ToggleAction, uid, actions);
+    }
+
+	protected override void UserUnequippedGloves(EntityUid uid, SpaceNinjaGlovesComponent comp, EntityUid user)
+	{
+		base.UserUnequippedGloves(uid, comp, user);
+
+        _actions.RemoveProvidedActions(user, uid);
+    }
+
+	// TODO: make this per-ability
+    private void OnActivate(ActivateInWorldEvent args)
+    {
+        var user = args.User;
+        var target = args.Target;
+        if (args.Handled
+            || !TryComp<SpaceNinjaComponent>(user, out var ninja)
+            || ninja.Gloves == null
+            || !HasComp<GlovesEnabledComponent>(ninja.Gloves))
+            return;
```

Comment:
You should swap this with AfterInteractEvent and have it connected with the Ninja Gloves component. AfterInteract is more for clicking where ActivateInWorld is more for using it in hand. Check out AirlockPainterSystem to see an example.

Reply:
but with airlock painter the painter has to be in hand, not equipped

wouldnt that make it require holding gloves to use them?

Reply:
> but with airlock painter the painter has to be in hand, not equipped
> 
> wouldnt that make it require holding gloves to use them?

Nope!

Reply:
didn't need to relay anything it just works
the one piece is real

Reply:
ok actually doesnt work, how would i relay the click
i hope its not client code with the funny events...

Reply:
I made some more review comments with the relay code.

Reply:
resolving in favour of the new review